### ISSUE TITLE
Problem with newline character(s) in different OSs

### ISSUE THREAD

**Issue by tsuburin** [markdownlint/markdownlint#237]
### Problem

MD009 and MD022 are incorrectly reported when a markdown is created in windows and then moved to linux.

### Environment

File creation: windows10
mdl execution: archlinux

### Reproduction

1. Create and save the following in a windows system.
```markdown
# foo

bar
```
2. Copy it to a linux system.
3. Run the following in the linux system (Let test.md be the name of the file).
```bash
mdl test.md
```
4. The following is output
```
test.md:1: MD009 Trailing spaces
test.md:2: MD009 Trailing spaces
test.md:3: MD009 Trailing spaces
test.md:1: MD022 Headers should be surrounded by blank lines
```

### Cause

The newline characters in windows are `\r\n` while that of linux is `\n`, so when splitting documents into lines,
https://github.com/markdownlint/markdownlint/blob/a9e80fcf3989d73b654b00bb2225a00be53983e8/lib/mdl/doc.rb#L44
`\r`s are left in those lines.

### Fix

Use `split(/\R/)` instead of `split("\n")`. #238

**Comment by jamesquilty** [comment]
This bug bit me hard, taking a few hours of my time to work out what was because the line endings are invisible _and_ the error thrown by mdl is a cryptic 'nil:NilClass (NoMethodError)'.

The PR is a one-line fix, which I've patched manually in my installed copy of mdl and can confirm that it works. It would be great to get this PR accepted. Who should be `@mentioned` for that?

**Edited to add:** `mdl` still chokes on a file with `0x0C` `^L` line termination characters even with this fix 🙁. The `0C` character appeared in two places in sequences like `0D 0D 0C 0D`. That said, that file is a complete mess, with just about every sequence of `0A` and `0D` you could imagine. No doubt a contributing factor to that problem is that _some_ of the original content was copy-and-pasted from Microsoft Word.

**Comment by apjanke** [comment]
0x0C is a Line Feed; that's an unusual one to see in the wild and I wouldn't expect markdownlint or Markdown itself to handle it well.

When you say "choke", does `mdl` actually throw an error, or does it report spurious rule violations?

Maybe the presence of Form Feed or other unusual cursor-positioning/layout-control characters should itself be a markdownlint rule violation?

**Comment by jamesquilty** [comment]
`0x0C` is a Form Feed `FF` and, yes, it's not at all common! I'm not sure where it's being introduced, it could be any of the software from Microsoft Word onward. When I say "choke" I mean that `mdl` halts with the same cryptic error message quoted above. In retrospect, this may well be a separate bug.

It would, indeed, be really great for `mdl` to handle control codes like `FF` a little more gracefully than the hard exit with a cryptic error (by which I mean, an error message originating from the Ruby code which doesn't make sense to an `mdl` user and the cause of which is hard to track down based on the message provided). Adding a rule for this would be really helpful, please!

**Comment by apjanke** [comment]
Ahh: I bet that `FF` is matching the regex for "whitespace" in the pattern used for rule `MD009 Trailing spaces` so it looks like trailing space at the end of a line, instead of a newline made with an unusual character. Maybe rule `MD009` should be made stricter about what it sees as "space", and cover the other unusual whitespace characters with a new separate rule that complains whenever they're present in the document at all.

Maybe the new rule should look for Form Feed, Vertical Tab, File Separator, Group Separator, and Record Separator, and MD009 should match any whitespace characters that are not in that set?

That `nil:NilClass (NoMethodError)` is definitely a bug; interesting that it's triggered here, but I don't think it's the same issue as the FF pattern matching. I'll try to create a minimal repro for this.

**Comment by apjanke** [comment]
Oh, I misunderstood: @jamesquilty, you were never getting those `MD009` and `MD022` rule messages; it was just bailing for you with the `NoMethodError`, right?

Any chance this is still happening for you and you can produce a stacktrace of where the NoMethodError is getting thrown? I can't reproduce this error with a simple document containing form feeds.

Or could you post the bogus Markdown file that's triggering this error as an attachment to this issue? You may have to zip it up so GitHub will let you attach it.

**Comment by jaymzh** [comment]
I agree, @apjanke, that FF it's a bit odd. Nonetheless, I think we should use `.lines` over `.split("\n")`, as I requested in the attached PR which should handle platforms better anyway.

**Comment by apjanke** [comment]
`.lines` seems like it would work here: I did some interactive testing, and it doesn't seem to consider FF to be a line separator character.

However, `.lines` leaves the line-ending characters in the returned strings:

```
irb(main):010:0> str = "foo\nbar\nbaz"
=> "foo\nbar\nbaz"
irb(main):011:0> str.lines
=> ["foo\n", "bar\n", "baz"]
irb(main):012:0> str = "foo\r\nbar\r\nbaz"
=> "foo\r\nbar\r\nbaz"
irb(main):013:0> str.lines
=> ["foo\r\n", "bar\r\n", "baz"]
```

So we'd still need to figure out a proper pattern for stripping those line endings while leaving other trailing whitespace intact, or adjust the rule definitions to account for their presence.

**Comment by jaymzh** [comment]
That's a good point... that's probably a more involved change.

**Comment by jamesquilty** [comment]
The bytestring is `41 0A 0C 0A 2A 20 42` will cause the problem for me with the `@lines = text.split(/\R/)` patch from #238 

```
A
^L
* B
```

The stack trace is 

```
$ mdl test.md
test.md:3: MD012 Multiple consecutive blank lines
/Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:473:in `block (4 levels) in load': undefined method `[]' for nil:NilClass (NoMethodError)
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:472:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:472:in `block (3 levels) in load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:466:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:466:in `block (2 levels) in load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:86:in `block (2 levels) in run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:84:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:84:in `block in run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:74:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:74:in `run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/bin/mdl:10:in `<top (required)>'
	from /Users/jamesquilty/.gem/ruby/2.3.0/bin/mdl:22:in `load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/bin/mdl:22:in `<main>'
```

Making the change in doc.rb to `@lines = text.lines` allows the above to pass, but `mdl` still dies on a bytestring `41 0D 0C 0D 2A 20 42` with

```
$ mdl test.md
/Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/doc.rb:173:in `header_style': undefined method `start_with?' for nil:NilClass (NoMethodError)
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:52:in `block (3 levels) in load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:51:in `map'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:51:in `block (2 levels) in load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:86:in `block (2 levels) in run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:84:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:84:in `block in run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:74:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:74:in `run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/bin/mdl:10:in `<top (required)>'
	from /Users/jamesquilty/.gem/ruby/2.3.0/bin/mdl:22:in `load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/bin/mdl:22:in `<main>'
```

I've also seen the following for the bytestring `41 0D 0A 0C 0D 0A 42`

```
$ mdl ENGR\ 301\ Team\ Contract\ copy.md
/Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/doc.rb:173:in `header_style': undefined method `start_with?' for nil:NilClass (NoMethodError)
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:52:in `block (3 levels) in load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:51:in `map'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/rules.rb:51:in `block (2 levels) in load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:86:in `block (2 levels) in run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:84:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:84:in `block in run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:74:in `each'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:74:in `run'
	from /Users/jamesquilty/.gem/ruby/2.3.0/gems/mdl-0.5.0/bin/mdl:10:in `<top (required)>'
	from /Users/jamesquilty/.gem/ruby/2.3.0/bin/mdl:22:in `load'
	from /Users/jamesquilty/.gem/ruby/2.3.0/bin/mdl:22:in `<main>'
```

So, it seems to be a very wide problem.

**Comment by DavidAnson** [comment]
For context, here is what the ConmonMark specification has to say about line-endings: https://spec.commonmark.org/0.29/#characters-and-lines

Note that there was a change as recently as 0.29!

**Comment by apjanke** [comment]
Thanks for the detailed info, @jamesquilty! Yeah, it doesn't look like those errors should be happening, regardless of what's in the input file.

And the [GitHub Flavored Markdown spec](https://github.github.com/gfm/#characters-and-lines) copies CommonMark verbatim. I say we orient ourselves to what CommonMark specifies.

**Comment by jaymzh** [comment]
So I wrote this script to generate test files for me:

```ruby
byte_strings = [
  "\x41\x0A\x0C\x0A\x2A\x20\x42",
  "\x41\x0D\x0C\x0D\x2A\x20\x42",
  "\x41\x0D\x0A\x0C\x0D\x0A\x42",
]

byte_strings.each_with_index do |bytes, index|
  file = "/tmp/wonk#{index + 1}"
  puts "Generating #{file}"
  open(file, 'w+b') do |fh|
    fh.write(bytes)
  end
end
```

I find that mdl, as it is, works fine on the 1st and 3rd strings and only crashes on the 2nd. That's on `master` with no modifications at all. Changing the split to use lines doesn't change this at all.

So the reason this fails is because once it parses it the "lines" it gets us:

```
["A\r\f\r* B"]
```

So that is something non-printable followed by a `*` followed by more non-printable. So **then** we get down to line list rule which tries to look the line number for this element and can't find it.

So I suspect the proper fix, might be something like this:

```
@@ -190,7 +191,12 @@ module MarkdownLint
       if item.type != :li
         raise "list_style called with non-list element"
       end
-      line = element_line(item).strip
+      line = element_line(item)
+      unless line
+        raise "list_style called with bogus data"
+      end
+
+      line.strip!
```

Which will still bail out, but gives a slightly clear error. Or possibly:

```
@@ -190,7 +191,14 @@ module MarkdownLint
       if item.type != :li
         raise "list_style called with non-list element"
       end
-      line = element_line(item).strip
+      line = element_line(item)
+      unless line
+        puts "Warning: Skipping potential list item, could not determine" +
+          " line: #{item}"
+        return :unknown
+      end
+      line.strip!
+
```

Which gets us further, but we then die trying to determine the indentation on the list a bit later. We can fake that out:

```
@@ -211,6 +219,7 @@ module MarkdownLint
     # indent of 8 spaces. You need to pass in the raw string here.
 
     def indent_for(line)
+      return 0 unless line
       return line.match(/^\s*/)[0].gsub("\t", " " * 8).length
     end
```

but we then crash later, and you see where that goes. I would argue the right solution is simply to throw a nicer error as-above.

**Comment by jaymzh** [comment]
I'm approving the associated PR, which solves the OP's platform issue... this unusual-byte-sequences issue is a different bug and probably needs a separate issue.

**Comment by jaymzh** [comment]
Let's move discussion to #278 about the odd byte sequences. Also - the attached PR actually will cause byte sequence 1 above to cause mdl to crash where as the current code doesn't, but I still think it's better to be windows-compatible in the majority of cases then prevent a crash in a corner case...

**Comment by jaymzh** [comment]
OK with the merging of #238, I'm marking this as fixed and waiting release, as the multi-OS stuff should be good.

**Comment by jaymzh** [comment]
This is fixed in 0.8.0 which is now available.
